### ISSUE TITLE
checkpoint(function, *args) should has the same requires_grad as function(*args)

### ISSUE THREAD

**Issue by ruotianluo** [pytorch/pytorch#7617]
## Issue description

In principle, to my understanding, checkpoint(function, *args) should behave exactly the same during forwarding as function(*args).

However, now, if the all the inputs have requires_grad to be False, the output will also checkpoint(...) will also have requires_grad to be False. While function may involve other trainable parameters, function(*args) will have requires_grad True output even all the inputs are requires_grad False.

## Code example

```
import torch
from torch.utils.checkpoint import checkpoint

m = torch.nn.Linear(4,3)
x = torch.randn(10,4)

z1 = checkpoint(lambda _:m(_), x)
print(z1.requires_grad)

z2 = m(x)
print(z2.requires_grad)
```

## System Info

PyTorch version: 0.5.0a0+a61d4a3
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.13.1
GCC version: Could not collect
CMake version: version 3.9.4

Python version: 2.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] numpy (1.14.2)
[pip] torch (0.5.0a0+a61d4a3, /Users/ruotianluo/anaconda/lib/python2.7/site-packages)
[pip] torchvision (0.2.0)
[conda] torch                     0.5.0a0+242f6c3           <pip>
[conda] torch                     0.5.0a0+a61d4a3           <pip>
[conda] torch                     0.4.0a0+dd91d57           <pip>
[conda] torch                     0.4.0a0+4a9e02f           <pip>
[conda] torch                     0.4.0a0+460e8cd           <pip>
[conda] torch                     0.4.0a0+4dd29ac           <pip>
[conda] torchvision               0.2.0                     <pip>

**Comment by apaszke** [comment]
cc @prigoyal @ezyang 

This can be observed exactly because we don't control the implicitly captured variables (as I mentioned previously).

**Comment by YoelShoshan** [comment]
I have encountered this issue just now.
In my scenario, the first thing that happens in my model is calling checkpoint few times using several feature extractors.
Is there any recommended workaround for this right now for users?

At first I thought about doing something like:

`input_data.requires_grad_(True)`

and then do:
`y = checkpoint(foo, input_data)`

But I think that it will also treat the input as parameters (like done in sensitivity/saliency map creation), which isn't something I need.

Should I add/use some sort of a dummy function before calling checkpoint? 
any suggestions?

**Comment by keven425** [comment]
running into the same issue. Is there any update? Thanks.

**Comment by ezyang** [comment]
No, there hasn't been a fix for this; it's just been documented as a known deficiency in the current implementation.

One possible workaround is to pass in a dummy argument with `requires_grad=True` to induce `requires_grad` on the output. It might be possible to formalize this hack in the API itself.

**Comment by keven425** [comment]
Ok. Is there any side effect in doing this? E.g. increased gpu memory usage?

**Comment by ezyang** [comment]
I don't think this should result in tensors getting retained when they shouldn't.

**Comment by YoelShoshan** [comment]
Just wanted to share that I've been using the dummy tensor trick for a while now, and it works well, the dummy variable is disconnected from the calculation graph, and it is tiny, so I strongly doubt that there are any actual side effects.

It's just "messy", some API or implementation change is probably better...

My guess is that some of the people who do not read the documentation well enough, or just aren't aware enough of the implementation, simply add requires_grad to their input (for example, input image), instead of the dummy tensor, which should work, but wastes computation and memory.

**Comment by keven425** [comment]
@YoelShoshan I copied your example from [here](https://discuss.pytorch.org/t/checkpoint-with-no-grad-requiring-inputs-problem/19117/11). However, when I print the gradient with `.grad` on the convolution weights, I still get None. Is there anything extra that I need to do?

```
import torch
from torch import nn
from torchvision import models
from torch.utils.checkpoint import checkpoint

class ModuleWrapperIgnores2ndArg(nn.Module):
    def __init__(self, module):
        super().__init__()
        self.module = module

    def forward(self,x, dummy_arg=None):
        assert dummy_arg is not None
        x = self.module(x)
        return x

class MyModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.features = nn.Sequential(*list(models.resnet18(pretrained=False).children())[:5])
        self.fc1 = nn.Linear(200704, 2)
        self.dummy_tensor = torch.ones(1, dtype=torch.float32, requires_grad=True)
        self.module_wrapper = ModuleWrapperIgnores2ndArg(self.features)

    def forward(self, x):
        #x = checkpoint(self.features, x)
        x = checkpoint(self.module_wrapper,x,self.dummy_tensor)
        print(x.shape)
        x = x.view(x.size(0), -1)
        x = self.fc1(x)
        return x

model = MyModel().cuda()
x = torch.randn(1, 3, 224, 224).cuda()
output = model(x)
output.mean().backward()
print(model.features[0].weight.grad)
```

**Comment by keven425** [comment]
@ezyang using the "dummy variable" method, I get gradient as `None`, if there is an `nn.InstanceNorm2d()`, followed by `nn.ReLU(inplace=True)`. If either line is removed, or if `Relu` is used with `inplace` set to `False`, the gradient is defined. 

Here is a minimal script to reproduce, based on @YoelShoshan 's work:


```
import torch
from torch import nn
from torch.utils.checkpoint import checkpoint

class ModuleWrapperIgnores2ndArg(nn.Module):
    def __init__(self, module):
        super().__init__()
        self.module = module

    def forward(self,x, dummy_arg=None):
        assert dummy_arg is not None
        x = self.module(x)
        return x

class MyModel(nn.Module):
    def __init__(self):
        super().__init__()
        in_ch = 3
        out_ch = 64
        self.features = nn.Sequential(
            nn.Conv2d(in_ch, out_ch, 3, padding=1),
            nn.InstanceNorm2d(out_ch),  # if this line is removed, works fine (grad is not None)
            nn.ReLU(inplace=True)       # if this line is removed, works fine (grad is not None)
        )
        self.fc1 = nn.Linear(3211264, 2)
        self.dummy_tensor = torch.ones(1, dtype=torch.float32, requires_grad=True)
        self.module_wrapper = ModuleWrapperIgnores2ndArg(self.features)

    def forward(self, x):
        x = checkpoint(self.module_wrapper,x,self.dummy_tensor)
        print(x.shape)
        x = x.view(x.size(0), -1)
        x = self.fc1(x)
        return x

model = MyModel().cuda()
x = torch.randn(1, 3, 224, 224).cuda()
output = model(x)
output.mean().backward()
print(model.features[0].weight.grad)
```

**Comment by alsrgv** [comment]
I'm having the same issue with `InstanceNorm2d`, `ReLU(inplace=True)` and gradient checkpointing.

**Comment by nyngwang** [comment]
@keven425 Did you solve it? I mean do you find any better workaround to facilitate this ducking API?

**Comment by nyngwang** [comment]
> One possible workaround is [..]

@ezyang If I use this workaround, what would be changed?

- The warning would disappear, only that.
- The warning would disappear, and the model wrapped by `checkpoint` would become trainable

Which one is true?

**Comment by nyngwang** [comment]
@albanD OK, it's my pleasure to try the brand new implementation :) Thanks for your reply.

**Comment by nyngwang** [comment]
@albanD How do I know using this API would really save GPU memory usage? Do you know how to profile the real GPU memory usage?

**Comment by albanD** [comment]
In both case, they should save similar amount of memory.
You might want to use nightly build to get the latest fixes though.

`nvidia-smi` is not a good way to measure memory as PyTorch will not release it to the main memory as soon as it is not used.
You can use `torch.cuda.memory_allocated()` to check how much memory is current being used.

**Comment by nyngwang** [comment]
To @albanD:

> You can use `torch.cuda.memory_allocated()` to check how much memory is current being used.

Thanks for the hint!

> You might want to use nightly build to get the latest fixes though.

Could you elaborate more on how to tune in to the latest information about this API? I can(and should) follow that by myself. That is: I would expect that there would be a thread somewhere(or on GitHub) about fixing(assumed that there is really a bug about not saving GPU memory).

Regarding the nightly build, how frequently does it get updated?

**Comment by nyngwang** [comment]
> @YoelShoshan Just wanted to share that I've been using the dummy tensor trick for a while now, and it works well,
> [...]
> It's just "messy", some API or implementation change is probably better...

For future readers: you can add this line before applying the `checkpoint` wrapper in the `forward(self, x)` method, so the messy class-version `ModuleWrapperIgnores2ndArg`(as shown above by @keven425) can be replaced with one line:

```
def forward(self, x):
  # ...
  x = x + torch.zeros(1, dtype=x.dtype, device=x.device, requires_grad=True)
  return checkpoint(..., use_reentrant=True) # it is True by default but I just make it explicit for emphasis.
```

**Comment by nyngwang** [comment]
@iamfaith You can take a look on my comment above. In short, you will need to set `use_reentrant=False` to avoid those buggy sh*t(even though I have found a simplified way, i.e. adding-one-line work, to resolve it) and save you a lot of time. But there is still one question in my mind: I don't know how to confirm that their implementation (in either ways, since they have written test to make sure the outputs from both are actually the same) does meet the paper [Training Deep Nets with Sublinear Memory Cost](https://arxiv.org/abs/1604.06174).

**Comment by albanD** [comment]
The tool provided here is rather simple and will only take care of dropping activation and recomputing for the function that is being wrapped.
The final effect will depend on which functions you wrap and at which granularity. Unfortunately, there isn't a way to know this ahead of time in the eager world of PyTorch so that's why this is the user's responsibility to do it.

**Comment by nyngwang** [comment]
@albanD Is there any new way to achieve the same as `use_reentrant=False` on Pytorch 2.0?

**Comment by ezyang** [comment]
Cc @wconstab
